implement Gradient

fujiisoup opened the pull request:
- [x] Closes #1332
 - [x] Tests added
 - [x] Tests passed
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

Added `xr.gradient`, `xr.DataArray.gradient`, and `xr.Dataset.gradient` according to #1332.

jakirkham commented on xarray/core/dask_array_ops.py:
```diff
@@ -104,3 +104,49 @@ def func(x, window, axis=-1):
     return out[index]
+
+
+def gradient(a, coord, axis, edge_order):
```
Do you mean like [this]( http://dask.pydata.org/en/latest/array-api.html#dask.array.gradient )? Requires Dask 0.17.3+ though. So IDK if that works for you or not.

fujiisoup replied:
Thanks @jakirkahm.
I need to compute the gradient with a non uniform coordinate, but dask Array.gradient only supports uniformly spacing coordinate.

I think this can be implemented using overlap as I do here. Is it in the scope of the dask development?

jakirkham replied:
Ah sorry. Should have read more of the code.

Yeah I think that is in scope. Just [not currently supported]( https://github.com/dask/dask/blob/5fa703eaaa6679e452c1dc7cdc8c9ac039994de5/dask/array/routines.py#L437-L438 ) yet.

Could you please open an issue or perhaps PR over on the Dask repo?

fujiisoup commented on xarray/core/computation.py:
```diff
@@ -1137,3 +1137,109 @@ def where(cond, x, y):
+      * x        (x) float64 0.0 0.1 1.1 1.2
+    Dimensions without coordinates: y
+    >>>
+    >>> xr.gradient(da, ('x', 'y'))
```
It would be another option, but in order to store them to a dataset, we will need to define their names very heuristically...

Maybe we can drop this api as this does not speed up the computation and it's easy to do the same thing manually.

shoyer replied:
I agree. I don't think it's particularly useful to compute to compute independent gradients with respect to multiple axes at the same time, e.g., `du/dx`, `du/dy`, etc. If anything, I would be more excited about computing higher order derivatives, e.g., `\frac{d^2 u}{dx dy}`

shoyer commented:
I think this will be very welcome functionality!

I wonder if we should consider calling this `xarray.differentiate` instead of `xarray.gradient`. I think the NumPy function is poorly named for differentiating along a single axis at once.

fujiisoup commented:
> I wonder if we should consider calling this xarray.differentiate instead of xarray.gradient. I think the NumPy function is poorly named for differentiating along a single axis at once.

Agreed. Differentiate is nicer.

Some other api questions arised during the implementation
+ Do we support `differentiate` for `Dataset`? In that case, what should we do for the variables that are independent from the target coordinate?
I thought 'keep them as is' is intuitive (and I implemented so), but mathematically, they should be zero.

+ Do we need to sort the array before computing `differentiate`? `np.gradient` implicitly assumes the array is sorted (but do nothing about this).

fujiisoup commented:
any thoughts for this?

fujiisoup commented:
Thanks, @dopplershift .

> Aren't you taking differences of values and dividing by differences between the matching coordinates?

Yes, correct. But if we have closer data points, then the estimate of the gradient becomes more precise. Sorting the array according to the coordinate provides the closest points, resulting in the most precise estimate of the gradient.

But I also think users can do it manually before taking the gradient.

shoyer commented on xarray/core/computation.py:
```diff
@@ -1137,3 +1137,81 @@ def where(cond, x, y):
+
+    if not isinstance(dataarray, DataArray):
+        raise TypeError(
+            'Only DataArray is supported. Given {}.'.format(type(dataarray)))
```
I'm a little confused here. You wrote an implementation for Dataset.differentiate, too, and here is a duplicate version of DataArray.differentiate.

shoyer replied:
👍 I don't think we need the separate function.

dcherian commented on xarray/core/dataarray.py:
```diff
@@ -2289,6 +2292,53 @@ def rank(self, dim, pct=False, keep_attrs=False):
         ds = self._to_temp_dataset().rank(dim, pct=pct, keep_attrs=keep_attrs)
         return self._from_temp_dataset(ds)
 
+    def differentiate(self, coord, edge_order=1):
```
Shouldn't this be `dim` instead of `coord`? I'm thinking of multidimensional co-ordinates here.

fujiisoup replied:
Currently, a 1-dimensional (non-dimensional) coordinate is also allowed. Do you think it is confusing?
For a multidimensional coordinate, ValueError will be raised.

dcherian replied:
OK¸ maybe add a little more to the docstring then to make it clear that we expect 1D co-ordinates.

dcherian commented on xarray/tests/test_dataset.py:
```diff
@@ -4492,3 +4492,38 @@ def test_raise_no_warning_for_nan_in_binary_ops():
+
+@pytest.mark.parametrize('dask', [True, False])
+@pytest.mark.parametrize('edge_order', [1, 2])
+def test_gradient(dask, edge_order):
```
can you add a test for a `datetime` co-ordinate also? The `gradient` in `npcompat` seems to support it.

fujiisoup replied:
It looks np.gradient does not support `datetime`. Maybe we need a unit that should be used for compute the gradient along `datetime`, such as `ns` or `date`. 
Do you have any idea about API for this?

dcherian replied:
Oh you're right; I misunderstood the code.

How about

```
da = xr.DataArray([1., 2.], dims='time', coords={'time': pd.date_range('2001-01-01', '2001-01-02')})

da.differentiate(coord='time', astype='D')
```

which would be equivalent to
```
np.diff(da.values) / (np.diff(da.time).astype('timedelta64[D]').astype('float64'))
```

`astype` isn't the best kwarg name though. `freq` or `resolution` don't seem correct either.

shoyer replied:
In the long term, hopefully we'll solve this by getting real unit support working, but a `time_unit` argument seems like a good choice for now

fujiisoup replied:
Numpy seems not to define the terminology for this unit, but *datetime_unit* is being used [here](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.datetime_data.html). I wonder if we should use it here.
Is it a little long?

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -2289,6 +2292,54 @@ def rank(self, dim, pct=False, keep_attrs=False):
+            The coordinate to be used to compute the gradient.
+        edge_order: 1 or 2. Default 1
+            N-th order accurate differences at the boundaries.
+
```
needs `time_unit` here

shoyer commented on xarray/core/dataset.py:
```diff
@@ -3663,6 +3666,60 @@ def rank(self, dim, pct=False, keep_attrs=False):
+        coord_data = coord_var.data
+        if coord_data.dtype.kind in ['m', 'M']:
+            if time_unit is None:
+                time_unit = np.datetime_data(coord_data.dtype)[0]
```
Use tuple unpacking here, e.g., `time_unit, _ = np.datetime_data(coord_data.dtype)`

fujiisoup commented:
I think it's ready :)

fmaussion commented:
This is so great! This is going to simplify my classes about derivatives on the globe even more: my only concern is that students will soon forget what a numerical derivative actually is, and that it's not a trivial to implement ;-)

shoyer commented:
Speaking of derivatives on the globe, we might want to include an option for periodic boundary conditions (and possibly for other functions like `interp` as well). But we can save that for another PR :).

fmaussion commented:
> But we can save that for another PR :)

See also https://github.com/pydata/xarray/issues/1288 : ``integrate`` is the next on my list ;) - I can try to give it a go if @fujiisoup doesn't want to do it himself

jakirkham commented:
@shoyer, have you seen `da.pad`?

shoyer commented:
> @shoyer, have you seen da.pad?

Yes, `pad` solves the hard part of periodic boundary conditions -- we should just need to do a bit of book-keeping on the xarray side.

rabernat commented:
I agree that the features implemented here are broadly useful and belong in xarray. The fundamental question is: how far does xarray itself want to go in supporting vector calculus in curvilinear coordinates (i.e. on the sphere). 

There is a fair bit of overlap between this new functionality and some of the things that we are trying to do in xgcm: https://xgcm.readthedocs.io/en/latest/grids.html. Xgcm supports periodic boundary conditions, as well as more complex topological connections between array edges. It allows users to reproduce precisely the sort of operations used to take gradients in finite-volume staggered-grid models.

shoyer commented:
@rabernat my inclination would be to draw the line at regular grids (with or without periodic boundary conditions), which are pretty commonly encountered in data analysis for any continuous physical systems. We would leave non-regular cases like staggered-grids to add-ons like xgcm -- these grids tend to be more application/PDE specific.

shoyer commented:
We should definitely mention specialized tools like xgcm as alternatives in the docstring for these xarray methods.

fujiisoup commented:
Thanks all. Updated.

Thanks, @spencerkclark 

> Eventually it would be nice if this worked on DataArrays with cftime.datetime coordinates; I think it would be relatively straightforward to modify to_numeric to enable it (we could probably enable it for interp at the same time), but I can take care of that later if you'd like.

Thanks. I added this function for something like this extension, though I do not yet fully follow your cftime update. It would be super nice if you could take care of this after merge.

@shoyer ,

> we might want to include an option for periodic boundary conditions 

Agreed. This option is nice not only `differentiate` but also `interp` and [`rolling`](#2011).
I think we can add a common logic to take care of them.

@fmaussion 

> See also #1288 : integrate is the next on my list ;) - I can try to give it a go if @fujiisoup doesn't want to do it himself

Thanks. 
Actually, I am now moving to another nation and would not have enough time in a few weeks.
I will appreciate if you could take care of this.

shoyer commented:
I'm going to merge this after the Appveyor tests pass, unless there are any further objections.

shoyer commented:
@rabernat let me know if this makes sense to you / if you agree.

rabernat commented on doc/computation.rst:
```diff
@@ -200,6 +200,26 @@ You can also use ``construct`` to compute a weighted rolling sum:
+                     coords=[0.1, 0.11, 0.2, 0.3])
+    a.differentiate('x')
+
+
```
This is a place where we could mention the limitations of `differentiate` and `gradient` for non-cartesian geometry.

rabernat commented on xarray/core/dataarray.py:
```diff
@@ -2289,6 +2292,57 @@ def rank(self, dim, pct=False, keep_attrs=False):
 
+    def differentiate(self, coord, edge_order=1, datetime_unit=None):
+        """ Differentiate the array with the second order accurate central
+        differences.
```
and here

rabernat commented on xarray/core/dataset.py:
```diff
@@ -3663,6 +3666,58 @@ def rank(self, dim, pct=False, keep_attrs=False):
 
+    def differentiate(self, coord, edge_order=1, datetime_unit=None):
+        """ Differentiate with the second order accurate central
+        differences.
```
and here

fujiisoup commented:
Thanks, @rabernat for the review.

Added the limitation of this method to docs.
